core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -207,4 +213,88 @@ private static boolean isOriginalPtfClass(Class<?> clazz, boolean includeDepreca
     }
+
+    private static class MethodPrinter
+    {
+        private static final int RELEVANT_METHOD_MODIFIERS = Modifier.PUBLIC | Modifier.PROTECTED | Modifier.PRIVATE | Modifier.STATIC;
+        private final Class<?> clazz;
+        private final Method method;
+
+        private MethodPrinter(Class<?> clazz, Method method)
+        {
+            this.clazz = clazz;
+            this.method = method;
+        }
+
+        // Based on java.lang.reflect.Executable.sharedToGenericString
+        // Two changes. Outputs only modifiers relevant to the compatibility check
+        // and uses specified class name instead of class where the method was declared in.
+        public String toString()
+        {
+            StringBuilder sb = new StringBuilder();
+
+            printModifiersIfNonzero(sb);
+
+            TypeVariable<?>[] typeparms = method.getTypeParameters();
+            if (typeparms.length > 0) {
+                sb.append(Arrays.stream(typeparms)
+                        .map(MethodPrinter::typeVarBounds)
+                        .collect(Collectors.joining(",", "<", "> ")));
+            }
+
+            specificToGenericStringHeader(sb);
+
+            sb.append('(');
+            StringJoiner sj = new StringJoiner(",");
+            Type[] params = method.getGenericParameterTypes();
+            for (int j = 0; j < params.length; j++) {
+                String param = params[j].getTypeName();
+                if (method.isVarArgs() && (j == params.length - 1)) { // replace T[] with T...
+                    param = param.replaceFirst("\\[\\]$", "...");
+                }
+                sj.add(param);
+            }
+            sb.append(sj);
+            sb.append(')');
+
+            Type[] exceptionTypes = method.getGenericExceptionTypes();
+            if (exceptionTypes.length > 0) {
+                sb.append(Arrays.stream(exceptionTypes)
+                        .map(Type::getTypeName)
+                        .collect(Collectors.joining(",", " throws ", "")));
+            }
+            return sb.toString();
+        }
+
+        void specificToGenericStringHeader(StringBuilder sb)
+        {
+            Type genRetType = method.getGenericReturnType();
+            sb.append(genRetType.getTypeName()).append(' ');
+            sb.append(clazz.getTypeName()).append('.');
+            sb.append(method.getName());
+        }
+
+        static String typeVarBounds(TypeVariable<?> typeVar)
+        {
+            Type[] bounds = typeVar.getBounds();
+            if (bounds.length == 1 && bounds[0].equals(Object.class)) {
+                return typeVar.getName();
+            }
+            else {
+                return typeVar.getName() + " extends " +
+                        Arrays.stream(bounds)
+                                .map(Type::getTypeName)
+                                .collect(Collectors.joining(" & "));
+            }
+        }
+
+        void printModifiersIfNonzero(StringBuilder sb)
```

Comment:
Maybe we can skip printing modifiers as we are only interested in public methods?